src/FastConv/FastConv.cpp

```diff
@@ -12,21 +13,298 @@ CFastConv::~CFastConv( void )
+
+    if (iLengthOfIr <= 0 || iBlockLength <= 0)
+        return Error_t::kFunctionInvalidArgsError;
+
+    switch (eCompMode)
+    {
+    case kTimeDomain:
+        m_pCCFastConvBase = new CFastConvTime(pfImpulseResponse, iLengthOfIr);
+        break;
+    case kFreqDomain:
+        m_pCCFastConvBase = new CFastConvFreq(pfImpulseResponse, iLengthOfIr, iBlockLength);
+        break;
+    default:
+        return Error_t::kFunctionInvalidArgsError;
+    }
 
     return Error_t::kNoError;
 }
 
 Error_t CFastConv::reset()
 {
+    if (m_pCCFastConvBase)
+    {
+        delete m_pCCFastConvBase;
+        m_pCCFastConvBase = 0;
+    }
     return Error_t::kNoError;
 }
 
-Error_t CFastConv::process (float* pfOutputBuffer, const float *pfInputBuffer, int iLengthOfBuffers )
+Error_t CFastConv::process(float* pfOutputBuffer, const float* pfInputBuffer, int iLengthOfBuffers)
 {
-    return Error_t::kNoError;
+    if (!pfOutputBuffer || !pfInputBuffer)
+        return Error_t::kMemError;
+
+    if (iLengthOfBuffers < 1 || iLengthOfBuffers > 1000000)
+        return Error_t::kFunctionInvalidArgsError;
+
+    return m_pCCFastConvBase->process(pfOutputBuffer, pfInputBuffer, iLengthOfBuffers);
 }
 
 Error_t CFastConv::flushBuffer(float* pfOutputBuffer)
 {
+    if (!pfOutputBuffer)
+        return Error_t::kMemError;
+
+    return m_pCCFastConvBase->flushBuffer(pfOutputBuffer);
+}
+int CFastConv::getTailLength() const
+{
+    return m_pCCFastConvBase->getTailLength();
+}
+//====================================================================
+
+//====================================================================
+CFastConvBase::CFastConvBase(float* pfIr, int iLengthOfIr)
+{
+    m_iLengthOfIr = iLengthOfIr;
+    m_pfIr = new float[m_iLengthOfIr];
+
+    CVectorFloat::copy(m_pfIr, pfIr, m_iLengthOfIr);
+}
+
+CFastConvBase::~CFastConvBase()
+{
+    delete[] m_pfIr;
+    m_pfIr = 0;
+
+    m_iLengthOfIr = 0;
+}
+
+Error_t CFastConvBase::flushBuffer(float* pfOutputBuffer)
+{
+    CVectorFloat::copy(pfOutputBuffer, m_pfTail, m_iLengthOfTail);
+    return Error_t::kNoError;
+}
+
+int CFastConvBase::getTailLength() const
+{
+    return m_iLengthOfTail;
+}
+//====================================================================
+
+//====================================================================
+CFastConvTime::CFastConvTime(float* pfIr, int iLengthOfIr) :
+    CFastConvBase(pfIr, iLengthOfIr)
+{
+    m_iLengthOfTail = iLengthOfIr - 1;
+    m_pfTail = new float[m_iLengthOfTail];
+    CVectorFloat::setZero(m_pfTail, m_iLengthOfTail);
+}
+
+CFastConvTime::~CFastConvTime()
+{
+    delete[] m_pfTail;
+    m_pfTail = 0;
+
+    m_iLengthOfTail = 0;
+}
+
+Error_t CFastConvTime::process(float* pfOutputBuffer, const float* pfInputBuffer, int iLengthOfBuffers)
+{
+
+    // Copy as many tail values as possible into output buffer
+    int iMinLength = std::min<int>(m_iLengthOfTail, iLengthOfBuffers);
+    CVectorFloat::copy(pfOutputBuffer, m_pfTail, iMinLength);
```

Comment:
This implementation for the reverb tail is really smart, you are avoiding one round of processing time by saving the output that is already pre-calculated compared to doing it as a separate process block operation with 0's as the input.